### ISSUE TITLE
[question] close real-time stream intentionally 

### ISSUE THREAD

**Issue by thangng48** [tardis-dev/tardis-node#45]
I've checked this https://github.com/tardis-dev/tardis-node/issues/30, but it seems not working properly.

Consider following code snippet, it basically re-subscribes some symbols to `deribit` exchange every minute. At the end of minute interval, it will break the combined stream --> expect to terminate underlying websockets to deribit. So, spinning new process will not flood the `deribit` server.
But, at the next iterations, `deribit` return `429` error code, that indicates the rate limit error.

@thaaddeus do we anyway to shutdown the stream or at least terminate the running websockets ?

```nodejs
const {
  combine,
  normalizeDerivativeTickers,
  normalizeLiquidations, normalizeOptionsSummary,
  normalizeTrades,
  streamNormalized
}  = require("tardis-dev");


async function* breakTick(end) {
  for (;!end.__break__;) {
    await new Promise((resolve) => setTimeout(resolve, 60000));
    yield {__break__: true}
  }
}

async function deribitStream() {
  const symbols = [
    "BTC-29JUL22-35000-C",
    "BTC-29JUL22-70000-C",
    "BTC-29JUL22-34000-P",
    "BTC-29JUL22-30000-C"
  ]

  const channels = ["trade", "derivative_ticker", "liquidation", "option_summary"]
  const end = {__break__: false}
  const streams = []

  for (const symbol of symbols){
    for (const channel of channels){
      const options = {
        exchange: "deribit",
        symbols: [symbol],
        withDisconnectMessages: true,
        onError: (err) => {
          console.log("Error subscribing to channel", {
            channel,
            exchange: "deribit",
            error: err,
          })
        }
      }
      let stream;
      switch (channel) {
        case "trade": {
          stream = streamNormalized(options, normalizeTrades)
          break
        }

        case "derivative_ticker": {
          stream = streamNormalized(options, normalizeDerivativeTickers)
          break
        }

        case "liquidation": {
          // Liquidations happen not frequently, disable feed timeout
          options.timeoutIntervalMS = 0

          stream = streamNormalized(options, normalizeLiquidations)
          break
        }

        case "option_summary": {
          stream = streamNormalized(options, normalizeOptionsSummary)
          break
        }
      }
      streams.push(stream)
    }
  }
  streams.push(breakTick(end))
  const stream = combine(...streams)

  for await (const message of stream){
    if (message.__break__){
      end.__break__ = true
      console.log("end stream")
      break
    }
    console.log(message)
  }
}

(async () => {
  for (;;){
    await deribitStream();
  }
})()
```

**Comment by thaaddeus** [comment]
best to add:

process.env.DEBUG = 'tardis-dev*'
at the beginning of the script so you'll be able to see if connections get closed or not.

**Comment by thaaddeus** [comment]
I can see the issue, unfortunately don't see an easy way to fix it, the issue is that code assumes that those individual streams push messages frequently and require next message to be pushed for a given connection to be closed.
This is unfortunate but with current design that is using async generators it's unsolvable right now, see https://github.com/tc39/proposal-async-iteration/issues/126
In practice what I'd suggest to do is to combine those low frequency streams into one stream/WS connection.

**Comment by thangng48** [comment]
@thaaddeus Can you add methods to allow disconnecting the underlying websocket directly ?

**Comment by kovetskiy** [comment]
@thaaddeus any updates on methods for intentional disconnects? Due to the lack of disconnect methods on the tardis-node side, an application that uses tardis-node can't gracefully shutdown when receiving such a signal because it can't be sure that the underlying ws connection is properly disconnected.

**Comment by thaaddeus** [comment]
I don't have any updates yet, pull requests welcome.

**Comment by lostless13** [comment]
Agreed with @kovetskiy. The ability to end streams gracefully would be useful. Looking through the code, the obvious approach is wrapping the _stream generator methods in a class is probably the correct approach to expose the close commands on the underlying websockets on the RealTimeStream classes. However, that change would cause a decent amount of breaking changes up to the library interface.

A hack might be to just keep a list of the open sockets in an array or object that is exposed to the library client allowing the client to kill the socket but that situation would still need to be handled by the generators.

Thanks!

**Comment by thaaddeus** [comment]
Fixed in `tardis-dev@18.0.0`. Stopping a combined managed real-time stream now promptly closes all underlying Tardis WebSocket sources, including quiet streams waiting on pending reads. We validated the original 16-stream Deribit pattern: all WebSockets returned to baseline in 46 ms with no errors or 429 responses. A custom async iterator may still delay completion of the outer `combine()` until its own pending `next()` settles, but the Tardis WebSockets are already closed.
